program-runtime/src/loaded_programs.rs

```diff
@@ -724,33 +722,17 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
                         Ordering::Relaxed,
                     );
                     self.stats.reloads.fetch_add(1, Ordering::Relaxed);
-                } else if existing.is_tombstone() != entry.is_tombstone() {
```

Comment:
How would a `FailedVerification` tombstone be replaced by a valid entry? For example:

1. A program gets deployed.
2. Environment gets updated and program fails to verify/compile. A tombstone is inserted.
3. Environment is updated again, and now the same program verified/compiles fine.

The deployment slot and effective slot hasn't changed for the program.

Reply:
Ack. And if we were able to remove this condition and tests didn't fail, we're missing tests for the new else branch now too?

Reply:
A program gets deployed => deployment slot changes
Environment gets updated => effective slot changes
So both are no replacement.

Reply:
Didn't realize that the effective slot is changing on environment update. It does make sense that it should change.
Where are we doing it in the code? I tried looking in bank.rs...

Reply:
https://github.com/solana-labs/solana/blob/864f29e93812c393ff5d221698cd72143a640cd0/runtime/src/bank.rs#L1364
https://github.com/solana-labs/solana/blob/864f29e93812c393ff5d221698cd72143a640cd0/svm/src/transaction_processor.rs#L744